cmd/bench/internal/client/client.go

```diff
@@ -0,0 +1,246 @@
+// stop returns true, at which point the goroutines quit and return the total
+// number of successfully completed requests. Some requests may be left
+// outstanding.
+func runBandwidthUntil(c BandwidthClient, stop func() bool) uint64 {
+	// We don't do any synchronization between the two goroutines, but the
+	// dispatch goroutine will eventually find stop() return false and will quit.
+	go func() {
+		for !stop() {
+			c.Dispatch()
+		}
+	}()
+
+	var count uint64
+	for !stop() {
+		c.Complete()
+		count++
+	}
+	return count
+}
+
+// A LatencyClient is a client used for latency measurements.
+type LatencyClient interface {
+	// Do executes a request and returns the duration.
+	Do() time.Duration
+}
+
+// FuncLatencyClient is a function that implements the LatencyClient interface.
+type FuncLatencyClient func() time.Duration
+
+// Do invokes f.
+func (f FuncLatencyClient) Do() time.Duration { return f() }
+
+var _ LatencyClient = FuncLatencyClient(nil)
+
+// RunLatencyClients runs the provided LatencyClients - each in its own
+// goroutine - for the duration d. It collects a histogram of the latencies
+// experienced by calls to Do. The lowest bucket will be from 0 to min, followed
+// by buckets whose width is given by step until max, followed by a bucket from
+// max to the maximum time.Duration.
+func RunLatencyClients(d time.Duration, min, max, step time.Duration, clients ...LatencyClient) []LatencyBucket {
+	if len(clients) == 0 {
+		panic("no clients")
+	}
+
+	// the marker will be atomically set to 1 when it's time for the goroutines to
+	// quit
+	var marker uint32
+	stop := func() bool { return atomic.LoadUint32(&marker) == 1 }
+
+	var histograms = make([][]LatencyBucket, len(clients))
+
+	var barrier, wg sync.WaitGroup
+	barrier.Add(1)
+	wg.Add(len(clients))
+	for i, c := range clients {
+		go func(i int, c LatencyClient) {
+			barrier.Wait()
+			histograms[i] = runLatencyUntil(c, min, max, step, stop)
+			wg.Done()
+		}(i, c)
+	}
+
+	barrier.Done()
+	time.Sleep(d)
+	atomic.StoreUint32(&marker, 1)
+	wg.Wait()
+
+	// since we got these histograms from runUntil, we know that they all have the
+	// same set and order of ranges, so we can trivially add them together to get
+	// the final answer. We arbitrarily pick histograms[0] to accumulate into.
+	buckets := histograms[0]
+	for _, bkts := range histograms[1:] {
+		for i, bkt := range bkts {
+			buckets[i].Count += bkt.Count
+		}
+	}
+	return buckets
+}
+
+func runLatencyUntil(c LatencyClient, bottom, top, skip time.Duration, stop func() bool) []LatencyBucket {
+	diff := top - bottom
+	nbuckets := int(diff / skip)
+	latencyCounts := make([]uint64, nbuckets+2)
+	for !stop() {
+		dur := c.Do()
+		if dur < bottom {
+			latencyCounts[0]++
+		} else if dur > top {
+			latencyCounts[nbuckets-1]++
+		} else {
+			bucketIdx := int((dur - bottom) / skip)
+			latencyCounts[bucketIdx]++
+		}
+	}
+
+	buckets := make([]LatencyBucket, nbuckets+2)
+	buckets[0].Until = bottom - 1
```

Comment:
The buckets are actually fully closed (see the comments on `LatencyBucket`) so that you can represent the maximum duration. It's kind of a stupid reason tbh, and I'd be happy to just make them half-open like you suggested and document that we don't support `time.Duration(math.MaxInt64)`.